File: common/src/main/java/com/ssblur/scriptor/registry/TokenGeneratorRegistry.java
package com.ssblur.scriptor.registry;

import com.google.gson.JsonObject;
import com.ssblur.scriptor.helpers.generators.CommunityModeGenerator;
import com.ssblur.scriptor.helpers.generators.MixedGroupGenerator;
import com.ssblur.scriptor.helpers.generators.StaticTokenGenerator;
import com.ssblur.scriptor.helpers.generators.TokenGenerator;
import net.minecraft.resources.ResourceLocation;

import javax.annotation.Nullable;
import java.util.HashMap;

public class TokenGeneratorRegistry {
  public static TokenGeneratorRegistry INSTANCE = new TokenGeneratorRegistry();

  public static final TokenGenerator.TokenGeneratorGenerator MIXED_GROUP =
    INSTANCE.registerGeneratorGenerator("mixed_groups", MixedGroupGenerator::new);
  public static final TokenGenerator.TokenGeneratorGenerator STATIC_TOKEN =
    INSTANCE.registerGeneratorGenerator("static_token", StaticTokenGenerator::new);
  public static final TokenGenerator.TokenGeneratorGenerator COMMUNITY =
    INSTANCE.registerGeneratorGenerator("community", CommunityModeGenerator::new);

  HashMap<ResourceLocation, TokenGenerator> generators = new HashMap<>();
  HashMap<String, ResourceLocation> generatorBindings = new HashMap<>();
  HashMap<String, JsonObject> generatorBindingConfig = new HashMap<>();
  HashMap<String, TokenGenerator.TokenGeneratorGenerator> generatorGenerators = new HashMap<>();
  ResourceLocation defaultGenerator;

  public TokenGenerator.TokenGeneratorGenerator registerGeneratorGenerator(String key, TokenGenerator.TokenGeneratorGenerator generatorGenerator) {
    generatorGenerators.put(key, generatorGenerator);
    return generatorGenerator;
  }

  public TokenGenerator.TokenGeneratorGenerator getGeneratorGenerator(String key) {
    return generatorGenerators.get(key);
  }

  public void registerGenerator(ResourceLocation key, TokenGenerator generator) {
    generators.put(key, generator);
  }

  public TokenGenerator getGenerator(ResourceLocation key) {
    return generators.get(key);
  }

  public void registerBinding(String word, ResourceLocation generator, @Nullable JsonObject parameters) {
    generatorBindings.put(word, generator);
    generatorBindingConfig.put(word, parameters);
  }

  public ResourceLocation getBinding(String word) {
    return generatorBindings.getOrDefault(word, defaultGenerator);
  }

  public void registerDefaultGenerator(ResourceLocation generator) {
    defaultGenerator = generator;
  }

  public ResourceLocation getDefaultGenerator() {
    return defaultGenerator;
  }

  public String generateWord(String word, @Nullable JsonObject object) {
    return getGenerator(getBinding(word)).generateToken(word, object);
  }

  public String generateWord(String word) {
    return generateWord(word, generatorBindingConfig.get(word));
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/registry/TokenGeneratorRegistry.kt
package com.ssblur.scriptor.registry

import com.google.gson.JsonObject
import com.ssblur.scriptor.helpers.generators.*
import com.ssblur.scriptor.helpers.generators.TokenGenerator.TokenGeneratorGenerator
import net.minecraft.resources.ResourceLocation

object TokenGeneratorRegistry {
    var generators: HashMap<ResourceLocation?, TokenGenerator> = HashMap()
    var generatorBindings: HashMap<String, ResourceLocation?> = HashMap()
    var generatorBindingConfig: HashMap<String, JsonObject?> = HashMap()
    var generatorGenerators: HashMap<String, TokenGeneratorGenerator> = HashMap()
    @JvmField
    var defaultGenerator: ResourceLocation? = null

    fun registerGeneratorGenerator(key: String, generatorGenerator: TokenGeneratorGenerator): TokenGeneratorGenerator {
        generatorGenerators[key] = generatorGenerator
        return generatorGenerator
    }

    fun getGeneratorGenerator(key: String): TokenGeneratorGenerator? {
        return generatorGenerators[key]
    }

    fun registerGenerator(key: ResourceLocation?, generator: TokenGenerator) {
        generators[key] = generator
    }

    fun getGenerator(key: ResourceLocation?): TokenGenerator? {
        return generators[key]
    }

    fun registerBinding(word: String, generator: ResourceLocation?, parameters: JsonObject?) {
        generatorBindings[word] = generator
        generatorBindingConfig[word] = parameters
    }

    fun getBinding(word: String): ResourceLocation? {
        return generatorBindings.getOrDefault(word, defaultGenerator)
    }

    fun registerDefaultGenerator(generator: ResourceLocation?) {
        defaultGenerator = generator
    }

    @JvmOverloads
    fun generateWord(word: String, `object`: JsonObject? = generatorBindingConfig[word]): String {
        return getGenerator(getBinding(word))!!.generateToken(word, `object`)
    }

    val MIXED_GROUP: TokenGeneratorGenerator = registerGeneratorGenerator("mixed_groups", TokenGeneratorGenerator { MixedGroupGenerator(it) })
    val STATIC_TOKEN: TokenGeneratorGenerator = registerGeneratorGenerator("static_token", TokenGeneratorGenerator { StaticTokenGenerator(it) })
    val COMMUNITY: TokenGeneratorGenerator = registerGeneratorGenerator("community", TokenGeneratorGenerator { CommunityModeGenerator(it) })
    val DEBUG: TokenGeneratorGenerator = registerGeneratorGenerator("debug", TokenGeneratorGenerator { DebugGenerator(it) })
}
